CPAC/nuisance/utils/compcor.py

```diff
@@ -181,9 +181,34 @@ def fallback_svd(a, full_matrices=True, compute_uv=True):
 
 
 def TR_string_to_float(tr):
-    if 'ms' in tr:
-        tr = float(tr.replace('ms', '')) / 1000
-    else:
-        tr = float(tr.replace('s', ''))
+    """
+    Convert TR string to seconds (float). Suffixes 's' or 'ms' to indicate
+    seconds or milliseconds.
+
+    Parameters
+    ----------
+    tr : TR string representation. May use suffixes 's' or 'ms' to indicate
+    seconds or milliseconds.
+
+    Returns
+    -------
+    tr in seconds (float)
+    """
+    if not isinstance(tr, str):
+        raise TypeError(f'Improper type for TR_string_to_float ({tr}).')
```

Comment:
I don't think the user needs to know that this specific function crashed; they just need to know that C-PAC needs repetition time and doesn't know it. I'm inclined to check sometime before this resource is set https://github.com/FCP-INDI/C-PAC/blob/44e2431785cdb953fd8aa53c8767f53c1dfeb810/CPAC/utils/utils.py#L974-L977 from the sidecar, but without digging in, I don't know if it's easy enough to know at that point if we'll need it.

If it's not easy to know if we need TR before we set it, I'd suggest something like
```suggestion
    if tr is None:
        raise TypeError('RepetitionTime is required but is unknown.')
```

Reply:
From a design perspective I would do that with a try catch where the function is used. Also either report the full stack trace for debugging purposes or a very user friendly message. But it's your call.

Reply:
I agree about reporting a stacktrace and a user-friendly message. My suggestion for trying sooner is if we can know early that it will definitely fail, we don't have to get to that function before we catch the issue

Reply:
For this and the discussion below I am not sure what to do - this function is used directly in the workflow [here](https://github.com/FCP-INDI/C-PAC/blob/653b8cd4198c3dda54b2eddf6f77350f3cfd0049/CPAC/nuisance/nuisance.py#L1273) and [here](https://github.com/FCP-INDI/C-PAC/blob/653b8cd4198c3dda54b2eddf6f77350f3cfd0049/CPAC/nuisance/nuisance.py#L1726) so I don't know which specific assumptions we can make.

Reply:
Yeah, I think ideally we'd refactor https://github.com/FCP-INDI/C-PAC/blob/173c9462c912001af09eb47a001312ed4a3e153d/CPAC/utils/datasource.py#L581-L637 https://github.com/FCP-INDI/C-PAC/blob/173c9462c912001af09eb47a001312ed4a3e153d/CPAC/utils/utils.py#L731-L989 into a NodeBlock so we only end up with `TR` and `tpattern` if we have them (instead of setting them to `None`) and then add a check at the top of NodeBlocks that require `TR`, but I think what you've done here is the next-best thing